<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>bam2gff_bed.pl</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_postfix@neige.apple.com" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#author">AUTHOR</a></li>
</ul>

-->


</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>bam2gff_bed.pl</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>bam2gff_bed.pl [--options...] &lt;filename&gt;</p>
<pre>

  Options:
  --in &lt;filename&gt;
  --bed | --gff
  --pe
  --type &lt;text&gt;
  --source &lt;text&gt;
  --out &lt;filename&gt; 
  --bigbed | --bb
  --bbapp &lt;/path/to/bedToBigBed&gt;
  --(no)gz
  --help</pre>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<p>The command line flags and descriptions:</p>
<dl>
<dt><strong><a name="in_filename" class="item">--in &lt;filename&gt;</a></strong>

<dd>
<p>Specify the file name of a .bam alignment file. The file should be indexed. 
If not, the program will attempt to automatically index it.</p>
</dd>
</li>
<dt><strong><a name="bed_gff" class="item">--bed | --gff</a></strong>

<dd>
<p>Specify the output file format. Either a BED file (6-columns) or GFF v.3 
file will be written. The default behavior is to write a BED file.</p>
</dd>
</li>
<dt><strong><a name="pe" class="item">--pe</a></strong>

<dd>
<p>Indicate that the BAM data is paired end, and that the generated feature 
should represent the insert fragment.</p>
</dd>
</li>
<dt><strong><a name="type_text" class="item">--type &lt;text&gt;</a></strong>

<dd>
<p>For GFF output files, specify the GFF type. The default is to use the 
input file base name, appended with either '_reads' or '_paired_reads'.</p>
</dd>
</li>
<dt><strong><a name="source_text" class="item">--source &lt;text&gt;</a></strong>

<dd>
<p>For GFF output files, specify the source tag. The default is 'Illumina'.</p>
</dd>
</li>
<dt><strong><a name="out" class="item">--out</a></strong>

<dd>
<p>Specify the output file name. The default for GFF output files is to use 
the GFF type; for BED output files the input file base name is used. A 
'.bed' extension is added if necessary.</p>
</dd>
</li>
<dt><strong><a name="bigbed" class="item">--bigbed</a></strong>

<dt><strong><a name="bb" class="item">--bb</a></strong>

<dd>
<p>When generating a BED formatted output file, optionally indicate that 
a binary BigBed file should be generated instead of a text BED file. 
A .bed file is first generated, then converted to a .bb file, and then 
the .bed file is removed.</p>
</dd>
</li>
<dt><strong><a name="bbapp_path_to_bedtobigbed" class="item">--bbapp &lt;/path/to/bedToBigBed&gt;</a></strong>

<dd>
<p>Specify the path to the Jim Kent's bedToBigBed conversion utility. The 
default is to first check the tim_db_helper.cfg configuration file for 
the application path. Failing that, it will search the default 
environment path for the utility. If found, it will automatically 
execute the utility to convert the bed file.</p>
</dd>
</li>
<dt><strong><a name="no_gz" class="item">--(no)gz</a></strong>

<dd>
<p>Specify whether (or not) the output file should be compressed with gzip.</p>
</dd>
</li>
<dt><strong><a name="help" class="item">--help</a></strong>

<dd>
<p>Display this POD documentation.</p>
</dd>
</li>
</dl>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This program will convert alignments in a BAM formatted file into either a
BED or GFF formatted feature file. For BED output, a 6-column BED file is
generated. For GFF output, a v.3 GFF file is generated. In both cases, the
alignment name, strand, and a score value is recorded.</p>
<p>Both single- and paired-end alignments may be converted. For single-end 
alignments, the generated score is the percent of matched bases (1-100).
For paired-end alignments, the genomic coordinates of the entire insert 
fragment is recorded. The length of the insert is recorded as the score.</p>
<p>For BED files, coordinates are adjusted to interbase format, according to 
the specification.</p>
<p>An option exists to further convert the BED file to an indexed, binary BigBed 
format. Jim Kent's bedToBigBed conversion utility must be available, and 
either a chromosome definition file or access to a Bio::DB database is required.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<pre>
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112</pre>
<p>This package is free software; you can redistribute it and/or modify
it under the terms of the GPL (either version 1, or at your option,
any later version) or the Artistic License 2.0.</p>

</body>

</html>
