<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>big_file2gff3.pl</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_postfix@neige.apple.com" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#author">AUTHOR</a></li>
</ul>

-->


</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>big_file2gff3.pl</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>big_file2gff3.pl [--options...] &lt;filename1.bw&gt; &lt;filename2.bb&gt; ...</p>
<pre>

  --in &lt;file&gt; or &lt;file1,file2,...&gt;
  --path &lt;/destination/path/for/bigfiles/&gt;
  --source &lt;text&gt;
  --name &lt;text&gt; or &lt;text1,text2,...&gt;
  --type &lt;text&gt;
  --strand [f|r|w|c|+|-|1|0|-1],...
  --rename
  --set
  --conf
  --help</pre>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<p>The command line flags and descriptions:</p>
<dl>
<dt><strong><a name="in_filename_bw_filename_bb_filename_bam" class="item">--in &lt;filename.bw | filename.bb | filename.bam&gt;</a></strong>

<dd>
<p>Provide the name(s) of the input BigFiles. Three types of files are 
supported: BigWig (.bw), BigBed (.bb), or BAM (.bam) files. 
The list may be specified three ways: by reiterating the 
--in option, providing a single comma-delimited list, or simply listing 
all files after the options (e.g. &quot;data_*.bw&quot;).</p>
</dd>
</li>
<dt><strong><a name="path_destination_path_for_bigfiles" class="item">--path &lt;/destination/path/for/bigfiles/&gt;</a></strong>

<dd>
<p>Provide the destination directory name for the BigFile files. If the
destination does not exist, then it will created. This directory should be
writeable by the user and readable by all (or at least the Apache and MySQL
users). If the input files are not currently located here, they will be
copied to the directory for you. Note that when generating a BigWigSet, a
unique directory for just the indicated files should be provided. The
default path is the current path for the input file.</p>
</dd>
</li>
<dt><strong><a name="source_text" class="item">--source &lt;text&gt;</a></strong>

<dd>
<p>Provide the name for the GFF feature's source for all of the input value. 
The default value is &quot;data&quot;. Unique values for each input file is not 
supported.</p>
</dd>
</li>
<dt><strong><a name="name_text" class="item">--name &lt;text&gt;</a></strong>

<dd>
<p>Provide the name(s) for the GFF feature(s). This will be used as the GFF 
feature's name. A unique name should be provided for each file, and may be 
specified as a single comma-delimited list or by reiterating the --name 
option. The default value to use the input file basename.</p>
</dd>
</li>
<dt><strong><a name="type_text" class="item">--type &lt;text&gt;</a></strong>

<dd>
<p>Provide the GFF type for the GFF features for all input files. By default, 
it re-uses the GFF name. Unique values for each input file is not supported.</p>
</dd>
</li>
<dt><strong><a name="strand_f_r_w_c_1_0_1" class="item">--strand [f|r|w|c|+|-|1|0|-1],...</a></strong>

<dd>
<p>Indicate which strand the feature will be located. Acceptable values include 
(f)orward, (r)everse, (w)atson, (c)rick, +, -, 1, or -1. By default no strand 
is used. For mulitple input files with different strands, use the option 
repeatedly or provide a comma-delimited list. If only one value is provided, 
it is used for all input files.</p>
</dd>
</li>
<dt><strong><a name="rename" class="item">--rename</a></strong>

<dd>
<p>Rename the input source file basenames as &quot;$source.$name&quot; when moving to the 
target directory. This may help in organization and clarity of file listings. 
Default is false.</p>
</dd>
</li>
<dt><strong><a name="set" class="item">--set</a></strong>

<dd>
<p>Indicate that all of the input BigWig files should be part of a BigWigSet,
which treats all of the BigWig files in the target directory as a single
database. This is primarily for GBrowse, as biotoolbox scripts (currently)
don't use the BigWigSet adaptor. A text file is written in the target
directory with metadata for each BigWig file (feature, source, strand,
name) as described in Bio::DB::BigWigSet documentation. Additional metadata 
may be manually added as necessary. The default is false.</p>
</dd>
</li>
<dt><strong><a name="conf" class="item">--conf</a></strong>

<dd>
<p>Write sample GBrowse database and track configuration stanzas. Each BigFile 
file will get individual stanzas, unless the --set option is enabled, where 
a single stanza for the BigWigSet is provided. This is helpful when setting 
up GBrowse database and configurations. Default is false.</p>
</dd>
</li>
<dt><strong><a name="help" class="item">--help</a></strong>

<dd>
<p>Display this POD help.</p>
</dd>
</li>
</dl>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This program will generate a GFF3 file with features representing a 
BigFile file. Two types of BigFiles are supported: BigWig and BigBed. BAM 
files are also supported. The generated features will encompass the entire 
length of each chromosome represented in the data file. The name of the data 
file and its absolute path is stored as a tag value in each feature. This 
tag value can be used by <strong>tim_db_helper.pm</strong> to collect data from the file 
with respect to various locations and features in the database.</p>
<p>The source data file is copied to the destination directory. The file may be 
renamed as &quot;source.name&quot; so as to avoid confusion when lots of files are 
dumped into the same directory.</p>
<p>The BigWig files may also be designated as a BigWigSet, with unique metadata 
assigned to each file (source, type, name, strand). The BigWigSet may be 
treated as a single database with multiple bigwig data sources by GBrowse. A 
metadata index file is written in the target directory as described in 
Bio::DB::BigWigSet.</p>
<p>Multiple source files may be designated, and each may have its own name. 
This facilitates multiple file processing.</p>
<p>The generated GFF3 file is written to the current directory. One GFF file is 
written for each input file, or one GFF file for a BigWigSet. It uses the 
provided GFF name as the basename for the file.</p>
<p>Optionally, sample database and track GBrowse configuration stanzas may also be 
written to the current directory to facilitate setting up GBrowse.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<pre>
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112</pre>
<p>This package is free software; you can redistribute it and/or modify
it under the terms of the GPL (either version 1, or at your option,
any later version) or the Artistic License 2.0.</p>

</body>

</html>
