<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>data2bed.pl</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_postfix@neige.apple.com" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#author">AUTHOR</a></li>
</ul>

-->


</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>data2bed.pl</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>data2bed.pl [--options...] &lt;filename&gt;</p>
<pre>

  Options:
  --in &lt;filename&gt;
  --ask
  --chr &lt;column_index&gt;
  --start &lt;column_index&gt;
  --stop | --end &lt;column_index&gt;
  --name &lt;column_index | base_text&gt;
  --score &lt;column_index&gt;
  --strand &lt;column_index&gt;
  --zero
  --out &lt;filename&gt; 
  --bigbed | --bb
  --chromof &lt;filename&gt;
  --db &lt;database&gt;
  --bbapp &lt;/path/to/bedToBigBed&gt;
  --(no)gz
  --help</pre>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<p>The command line flags and descriptions:</p>
<dl>
<dt><strong><a name="in_filename" class="item">--in &lt;filename&gt;</a></strong>

<dd>
<p>Specify the file name of a data file. It must be a tab-delimited text file,
preferably in the tim data format as described in 'tim_file_helper.pm', 
although any format should work. The file may be compressed with gzip.</p>
</dd>
</li>
<dt><strong><a name="ask" class="item">--ask</a></strong>

<dd>
<p>Indicate that the program should interactively ask for the indices for 
feature data. It will present a list of the column 
names to choose from. Enter nothing for non-relevant columns or to 
accept default values.</p>
</dd>
</li>
<dt><strong><a name="chr_column_index" class="item">--chr &lt;column_index&gt;</a></strong>

<dd>
<p>The index of the dataset in the data table to be used 
as the chromosome or sequence ID column in the BED data.</p>
</dd>
</li>
<dt><strong><a name="start_column_index" class="item">--start &lt;column_index&gt;</a></strong>

<dd>
<p>The index of the dataset in the data table to be used 
as the start position column in the BED data.</p>
</dd>
</li>
<dt><strong><a name="start_column_index2" class="item">--start &lt;column_index&gt;</a></strong>

<dt><strong><a name="end_column_index" class="item">--end &lt;column_index&gt;</a></strong>

<dd>
<p>The index of the dataset in the data table to be used 
as the stop or end position column in the BED data.</p>
</dd>
</li>
<dt><strong><a name="score_column_index" class="item">--score &lt;column_index&gt;</a></strong>

<dd>
<p>The index of the dataset in the data table to be used 
as the score column in the BED data.</p>
</dd>
</li>
<dt><strong><a name="name_column_index_base_text" class="item">--name &lt;column_index | base_text&gt;</a></strong>

<dd>
<p>Supply either the index of the column in the data table to 
be used as the name column in the BED data, or the base text 
to be used when auto-generating unique feature names. The 
auto-generated names are in the format 'text_00000001'.</p>
</dd>
</li>
<dt><strong><a name="strand_column_index" class="item">--strand &lt;column_index&gt;</a></strong>

<dd>
<p>The index of the dataset in the data table to be used
for strand information. Accepted values might include
any of the following &quot;f(orward), r(everse), w(atson),
c(rick), +, -, 1, -1, 0, .&quot;.</p>
</dd>
</li>
<dt><strong><a name="zero" class="item">--zero</a></strong>

<dd>
<p>Indicate that the source data is already in interbase (0-based) 
coordinates and do not need to be converted. By convention, all 
BioPerl (and, by extension, all biotoolbox) scripts are base 
(1-based) coordinates. Default behavior is to convert.</p>
</dd>
</li>
<dt><strong><a name="out_filename" class="item">--out &lt;filename&gt;</a></strong>

<dd>
<p>Specify the output filename. By default it uses the basename of the input 
file.</p>
</dd>
</li>
<dt><strong><a name="bigbed" class="item">--bigbed</a></strong>

<dt><strong><a name="bb" class="item">--bb</a></strong>

<dd>
<p>Indicate that a binary BigBed file should be generated instead of 
a text BED file. A .bed file is first generated, then converted to 
a .bb file, and then the .bed file is removed.</p>
</dd>
</li>
<dt><strong><a name="chromf_filename" class="item">--chromf &lt;filename&gt;</a></strong>

<dd>
<p>When converting to a BigBed file, provide a two-column tab-delimited 
text file containing the chromosome names and their lengths in bp. 
Alternatively, provide a name of a database, below.</p>
</dd>
</li>
<dt><strong><a name="db_database" class="item">--db &lt;database&gt;</a></strong>

<dd>
<p>Specify the database from which chromosome lengths can be derived when 
generating a BigBed file. This option is only required when generating 
BigBed files. It may also be supplied from the metadata in the source 
data file.</p>
</dd>
</li>
<dt><strong><a name="bbapp_path_to_bedtobigbed" class="item">--bbapp &lt;/path/to/bedToBigBed&gt;</a></strong>

<dd>
<p>Specify the path to the Jim Kent's bedToBigBed conversion utility. The 
default is to first check the tim_db_helper.cfg configuration file for 
the application path. Failing that, it will search the default 
environment path for the utility. If found, it will automatically 
execute the utility to convert the bed file.</p>
</dd>
</li>
<dt><strong><a name="no_gz" class="item">--(no)gz</a></strong>

<dd>
<p>Specify whether (or not) the output file should be compressed with gzip.</p>
</dd>
</li>
<dt><strong><a name="help" class="item">--help</a></strong>

<dd>
<p>Display this POD documentation.</p>
</dd>
</li>
</dl>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This program will convert a tab-delimited data file into a BED file,
according to the specifications here
<a href="http://genome.ucsc.edu/goldenPath/help/customTrack.html#BED.">http://genome.ucsc.edu/goldenPath/help/customTrack.html#BED.</a> A minimum of
three and a maximum of six columns may be generated. Column identification 
may be specified on the command line or interactively. GFF source files  
have columns automatically identified. All lower-numbered
columns must be defined before writing higher-numbered columns. Thin and
thick block data are not written. Browser and Track lines are also not
written. Following specification, all coordinates are written in interbase
(0-based) coordinates. Base (1-based) coordinates (the BioPerl standard) will
be converted. Score values are not converted, however, to a 1..1000 scale.
The biotoolbox script <code>manipulate_datasets.pl</code> has tools to do this if
required.</p>
<p>An option exists to further convert the BED file to an indexed, binary BigBed 
format. Jim Kent's bedToBigBed conversion utility must be available, and 
either a chromosome definition file or access to a Bio::DB database is required.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<pre>
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112</pre>
<p>This package is free software; you can redistribute it and/or modify
it under the terms of the GPL (either version 1, or at your option,
any later version) or the Artistic License 2.0.</p>

</body>

</html>
