<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>novo_wrapper.pl</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_postfix@neige.apple.com" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#author">AUTHOR</a></li>
</ul>

-->


</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>novo_wrapper.pl</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<pre>

 novo_wrapper.pl --index &lt;file&gt; [--options] &lt;seq_file1&gt; ...</pre>
<pre>

  Options:
  --in &lt;seq_file&gt;
  --index &lt;novo_index&gt;
  --novo &lt;/path/to/novoalign&gt;
  --sam &lt;/path/to/samtools&gt;
  --repeat [None|All|Random|Exhaustive]
  --(no)il
  --temp &lt;/path/to/temp/directory&gt;
  --help</pre>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<p>The command line flags and descriptions:</p>
<dl>
<dt><strong><a name="in_seq_file" class="item">--in &lt;seq_file&gt;</a></strong>

<dd>
<p>Specify the name(s) of the input sequence files. Multiple files may be 
specified using multiple --in arguments, with a single --in argument and a 
comma delimited list, or as a list at the end of all options. The files may 
kept gzipped; they will automatically be uncompressed.</p>
</dd>
</li>
<dt><strong><a name="index_novo_index" class="item">--index &lt;novo_index&gt;</a></strong>

<dd>
<p>Specify the indexed genome database file to which the sequences will be 
aligned. The file should be built using Novocraft's 'novoindex' program.</p>
</dd>
</li>
<dt><strong><a name="novo_path_to_novoalign" class="item">--novo &lt;/path/to/novoalign&gt;</a></strong>

<dd>
<p>Specify the full path to Novocraft's 'novoalign' executable file. This 
argument is not necessary if it is located in the path.</p>
</dd>
</li>
<dt><strong><a name="sam_path_to_samtools" class="item">--sam &lt;/path/to/samtools&gt;</a></strong>

<dd>
<p>Specify the full path to the 'samtools' executable file. This 
argument is not necessary if it is located in the path.</p>
</dd>
</li>
<dt><strong><a name="repeat_none_all_random_exhaustive" class="item">--repeat [None|All|Random|Exhaustive]</a></strong>

<dd>
<p>Specify the novoalign option of dealing with sequences which align to 
multiple locations in the genome (repeats). See the novoalign documenation 
for more information. Default value is None.</p>
</dd>
</li>
<dt><strong><a name="no_il" class="item">--(no)il</a></strong>

<dd>
<p>Indicate whether the input sequence file is (not) an Illumina sequence file 
from the Illumina Pipeline v. 1.3. See the novoalign documentation for more 
information. Default value is true.</p>
</dd>
</li>
<dt><strong><a name="temp_path_to_temp_directory" class="item">--temp &lt;/path/to/temp/directory&gt;</a></strong>

<dd>
<p>Specify a temp directory for writing uncompressed sequence files and 
intermediate files. By default it uses the temp directory specified in 
the user's environment.</p>
</dd>
</li>
<dt><strong><a name="help" class="item">--help</a></strong>

<dd>
<p>This help text.</p>
</dd>
</li>
</dl>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This program is a simple wrapper for the novoalign and samtools programs, and 
performs the same functions as a custom shell script, including aligning the 
sequence reads to the genome using Novacraft's novoalign and then converting 
the alignments to an indexed, sorted, binary bam file using samtools, but 
perhaps in a slightly more friendly, if bloated, script.</p>
<p>The temp directory is used for temporary uncompressed sequence files and 
intermediate files to avoid unnecessary scheduled backups and indexing.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<pre>
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112

</pre>
<p>This package is free software; you can redistribute it and/or modify
it under the terms of the GPL (either version 1, or at your option,
any later version) or the Artistic License 2.0.</p>

</body>

</html>
