<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>wig2data.pl</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_postfix@neige.apple.com" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>
<!--

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
</ul>

-->


</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>wig2gff.pl</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>wig2gff.pl [--options...] &lt;filename&gt;</p>
<pre>

  Options:
  --in &lt;filename&gt;
  --out &lt;filename&gt; 
  --(no)gff
  --type &lt;text&gt;
  --source &lt;text&gt;
  --format [0,1,2,3]
  --(no)midpoint
  --version [2,3]
  --(no)gz
  --help</pre>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<p>The command line flags and descriptions:</p>
<dl>
<dt><strong><a name="in_filename" class="item">--in &lt;filename&gt;</a></strong>

<dd>
<p>Specify the file name of a wig file. The file may be compressed with gzip.</p>
</dd>
</li>
<dt><strong><a name="out_filename" class="item">--out &lt;filename&gt;</a></strong>

<dd>
<p>Specify the output filename. By default it uses the GFF type as the 
basename.</p>
</dd>
</li>
<dt><strong><a name="no_gff" class="item">--(no)gff</a></strong>

<dd>
<p>Indicate whether the output file should be in GFF format. If false, a 
standard tim data tab delimited text file will be written with four 
columns: chromosome, start, stop, and score. The default value is false.</p>
</dd>
</li>
<dt><strong><a name="type_text" class="item">--type &lt;text&gt;</a></strong>

<dd>
<p>Specify the text string to be used as the GFF feature 'type' or 
'method' (the 3rd column). By default it uses the name specified in the 
track line; otherwise, it uses the basename of the input wig file.</p>
</dd>
</li>
<dt><strong><a name="source_text" class="item">--source &lt;text&gt;</a></strong>

<dd>
<p>Specify the text string to be used as the GFF feature 'source' 
(the 2nd column). The default value is none.</p>
</dd>
</li>
<dt><strong><a name="format_0_1_2_3" class="item">--format [0,1,2,3]</a></strong>

<dd>
<p>Specify the number of decimal places to which the wig file will be 
formatted. Default is no formatting.</p>
</dd>
</li>
<dt><strong><a name="no_midpoint" class="item">--(no)midpoint</a></strong>

<dd>
<p>Specify whether (or not) a midpoint position should be calculated 
between the start and stop positions and be used in the output GFF 
file. This only pertains to BED style wig files where both the 
start and stop positions are reported and stepped wig files where 
a span value is specified. The default value is false.</p>
</dd>
</li>
<dt><strong><a name="version_2_3" class="item">--version [2,3]</a></strong>

<dd>
<p>Specify the GFF version. The default is version 3.</p>
</dd>
</li>
<dt><strong><a name="no_gz" class="item">--(no)gz</a></strong>

<dd>
<p>Specify whether (or not) the output file should be compressed with gzip.</p>
</dd>
</li>
<dt><strong><a name="help" class="item">--help</a></strong>

<dd>
<p>Display this POD documentation.</p>
</dd>
</li>
</dl>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This program will convert a wiggle data file into a tabbed delimited text 
data file. The data file will have four columns: chromosome, start, stop, 
and score. Alternatively, a GFF file may be written, in which case the 
GFF source and type values should be specified.</p>
<p>Wiggle files are used with the UCSC Genome Browser and can have multiple
formats, including BED (also referred to as bedgraph, variable step, and
fixed step. This program can convert all three formats. Improperly
formatted wig files may cause the program to die. More information about
wig files may be obtained from here:
<a href="http://genome.ucsc.edu/goldenPath/help/wiggle.html">http://genome.ucsc.edu/goldenPath/help/wiggle.html</a></p>
<p>The start position will be the coordinate listed in the wig file. If a span
value is indicated in the wiggle file, the GFF stop will equal start plus
span; otherwise stop will equal the start value. The score value may be
formatted to the indicated number of decimal places.</p>
<pre>

=head1 AUTHOR</pre>
<pre>
 Timothy J. Parnell, PhD
 Howard Hughes Medical Institute
 Dept of Oncological Sciences
 Huntsman Cancer Institute
 University of Utah
 Salt Lake City, UT, 84112

</pre>
<p>This package is free software; you can redistribute it and/or modify
it under the terms of the GPL (either version 1, or at your option,
any later version) or the Artistic License 2.0.</p>

</body>

</html>
